core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -202,32 +234,11 @@ private void advanceBufferIfNecessary() {
   }
 
   /**
-   * A simple read operation to perform on a single {@link ReadableBuffer}. All state management for
-   * the buffers is done by {@link CompositeReadableBuffer#execute(ReadOperation, int)}.
+   * A simple read operation to perform on a single {@link ReadableBuffer}.
+   * All state management for the buffers is done by
+   * {@link CompositeReadableBuffer#execute(ReadOperation, int, Object, int)}.
    */
-  private abstract static class ReadOperation {
-    /**
-     * Only used by {@link CompositeReadableBuffer#readUnsignedByte()}.
-     */
-    int value;
-
-    /**
-     * Only used by {@link CompositeReadableBuffer#readBytes(OutputStream, int)}.
-     */
-    IOException ex;
-
-    final void read(ReadableBuffer buffer, int length) {
-      try {
-        value = readInternal(buffer, length);
-      } catch (IOException e) {
-        ex = e;
-      }
-    }
-
-    final boolean isError() {
-      return ex != null;
-    }
-
-    abstract int readInternal(ReadableBuffer buffer, int length) throws IOException;
+  private interface ReadOperation<T> {
```

Comment:
Don't know if we care, but we could have a `NonThrowingReadOperation` that extends this interface and doesn't have `throws IOException`. `executeNoThrow()` could then have that as a parameter. That would preserve the checked exception checking.

Reply:
I may not completely follow, but I don't see that this would buy much. Unless we wanted to also duplicate the `execute()` method contents which wouldn't be ideal.

Reply:
No need to duplicate execute(). `execute()` and `executeNoThrow()` would work as you have it here. The difference is that `executeNoThrow()` could only be called with a ReadOperation that actually doesn't throw IOExceptions.